[inductor][3/N] triton support post-#5512, tt.divisibility format

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #145515
* #145583
* __->__ #145575

1. Fix the tt.divisibility format in hints.py. Previously, it was `{((0,), (1,)): [["tt.divisibility", 16]]}`. Now it is `{(0,): [["tt.divisibility", 16]], (1,): [["tt.divisibility", 16]]}`. This was an oversight in the first PR I added. I've verified that we now get `{ tt.divisibility = 16 }` in the generated TTGIR.
2. Update the test_codegen_triton.py test to work with multiple triton versions (and test this divisibility format in the new triton version)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

davidberard98 committed:
[inductor][4/N] triton support post-#5512, tt.divisibility format

1. Fix the tt.divisibility format in hints.py. Previously, it was
   `{((0,), (1,)): [["tt.divisibility", 16]]}`.
   Now it is
   `{(0,): [["tt.divisibility", 16]], (1,): [["tt.divisibility", 16]]}`.
   This was an oversight in the first version. I've verified in the
   now get { tt.divisibility = 16 } in the generated TTGIR.
2. Update the test_codegen_triton.py test to work with multiple triton
   versions (and test this divisibility format in the new triton
   version)

[ghstack-poisoned]

davidberard98 commented:
@anmyachev I would appreciate if you can take a quick look at the tt.divisibility format change!

I copied the original format from your patch, but noticed that the format is a bit different in the version of triton I'm testing with.

davidberard98 committed:
Update on "[inductor][3/N] triton support post-#5512, tt.divisibility format"


1. Fix the tt.divisibility format in hints.py. Previously, it was `{((0,), (1,)): [["tt.divisibility", 16]]}`. Now it is `{(0,): [["tt.divisibility", 16]], (1,): [["tt.divisibility", 16]]}`. This was an oversight in the first PR I added. I've verified that we now get `{ tt.divisibility = 16 }` in the generated TTGIR.
2. Update the test_codegen_triton.py test to work with multiple triton versions (and test this divisibility format in the new triton version)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor][3/N] triton support post-#5512, tt.divisibility format"


1. Fix the tt.divisibility format in hints.py. Previously, it was `{((0,), (1,)): [["tt.divisibility", 16]]}`. Now it is `{(0,): [["tt.divisibility", 16]], (1,): [["tt.divisibility", 16]]}`. This was an oversight in the first PR I added. I've verified that we now get `{ tt.divisibility = 16 }` in the generated TTGIR.
2. Update the test_codegen_triton.py test to work with multiple triton versions (and test this divisibility format in the new triton version)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

anmyachev commented:
> @anmyachev I would appreciate if you could take a quick look at the tt.divisibility format change!
> 
> I copied the original format from your patch, but noticed that the format is a bit different in the version of triton I'm testing with.

Hi @davidberard98! 

Indeed, I can confirm that the format now looks like this: `{(0,): [('tt.divisibility', 16)], (1,): [('tt.divisibility', 16)], (2,): [('tt.divisibility', 16)], (8,): [('tt.divisibility', 16)], (10,): [('tt.divisibility', 16)]}` (from one of Triton AOT tests). The only small difference is that it is not a list within a list (`[[]]`), but a tuple within a list (`[()]`).

I must have missed this change in PR#5512. Thanks for letting me know and fixing it so quickly!

davidberard98 commented:
@anmyachev thanks for taking a look!

re: tuple vs. list: interesting, I see some `("tt.divisibility", 16)` tuples

https://github.com/triton-lang/triton/blob/ad16e3d908791327f7c09c410c9294af9b03cfa1/python/triton/tools/compile.py#L109

and some `["tt.divisibility", 16]` lists:

https://github.com/triton-lang/triton/blob/ad16e3d908791327f7c09c410c9294af9b03cfa1/python/triton/backends/compiler.py#L92

davidberard98 committed:
Update on "[inductor][3/N] triton support post-#5512, tt.divisibility format"


1. Fix the tt.divisibility format in hints.py. Previously, it was `{((0,), (1,)): [["tt.divisibility", 16]]}`. Now it is `{(0,): [["tt.divisibility", 16]], (1,): [["tt.divisibility", 16]]}`. This was an oversight in the first PR I added. I've verified that we now get `{ tt.divisibility = 16 }` in the generated TTGIR.
2. Update the test_codegen_triton.py test to work with multiple triton versions (and test this divisibility format in the new triton version)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

anmyachev commented:
> @anmyachev thanks for taking a look!
> 
> re: tuple vs. list: interesting, I see some `("tt.divisibility", 16)` tuples
> 
> https://github.com/triton-lang/triton/blob/ad16e3d908791327f7c09c410c9294af9b03cfa1/python/triton/tools/compile.py#L109
> 
> and some `["tt.divisibility", 16]` lists:
> 
> https://github.com/triton-lang/triton/blob/ad16e3d908791327f7c09c410c9294af9b03cfa1/python/triton/backends/compiler.py#L92

Now it should behave the same everywhere: https://github.com/triton-lang/triton/pull/5702

davidberard98 commented:
Thanks @anmyachev !

SamGinzburg commented on torch/_inductor/runtime/hints.py:
```diff
@@ -95,9 +95,7 @@ def AttrsDescriptorWrapper(
             divisible_by_16=None,
             equal_to_1=None,
         ):
-            return {
```
resolved offline, I think as long as we document this behavior (the ~Dec 9, 2024 to Jan 1 2025 version) and explicitly error on it at some point we should be good.

We will track this here: https://github.com/pytorch/pytorch/issues/144103 for now

SamGinzburg reviewed (approved):
Changes look good to me (discussed offline)

davidberard98 commented:
@pytorchbot merge
